executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,272 @@
+            Identity::Connect(i) => i.children(),
+            Identity::BusInteraction(i) => i.children(),
+        }
+    }
+}
+
+impl<T> Identity<T> {
+    pub fn contains_next_ref(&self) -> bool {
+        self.children().any(|e| e.contains_next_ref())
+    }
+
+    pub fn id(&self) -> u64 {
+        match self {
+            Identity::Polynomial(i) => i.id,
+            Identity::Connect(i) => i.id,
+            Identity::BusInteraction(i) => i.id,
+        }
+    }
+}
+
+pub fn convert<T: Clone>(identities: &[AnalyzedIdentity<T>]) -> Vec<Identity<T>> {
+    let max_poly_id = identities
+        .iter()
+        .map(|i| match i {
+            AnalyzedIdentity::Polynomial(identity) => identity.id,
+            AnalyzedIdentity::Connect(identity) => identity.id,
+            // Replaced anyway, so we don't bother with the ID.
+            AnalyzedIdentity::Lookup(_)
+            | AnalyzedIdentity::PhantomLookup(_)
+            | AnalyzedIdentity::Permutation(_)
+            | AnalyzedIdentity::PhantomPermutation(_)
+            | AnalyzedIdentity::PhantomBusInteraction(_) => 0,
+        })
+        .max()
+        .unwrap_or(0);
+
+    let mut id_counter = max_poly_id + 1;
+
+    identities
+        .iter()
+        .flat_map(|i| match i {
+            AnalyzedIdentity::Polynomial(identity) => {
+                vec![Identity::Polynomial(identity.clone())].into_iter()
+            }
+            AnalyzedIdentity::Connect(identity) => {
+                vec![Identity::Connect(identity.clone())].into_iter()
+            }
+            AnalyzedIdentity::PhantomBusInteraction(identity) => {
+                let id = id_counter;
+                id_counter += 1;
+                let negative_multiplicity = match &identity.multiplicity {
+                    AlgebraicExpression::UnaryOperation(op) => {
+                        op.op == AlgebraicUnaryOperator::Minus
+                    }
+                    _ => false,
+                };
+                let interaction_type = if negative_multiplicity {
+                    InteractionType::Receive
+                } else {
+                    InteractionType::Send
+                };
+                vec![Identity::BusInteraction(BusInteractionIdentity {
+                    id,
+                    multiplicity: Some(identity.multiplicity.clone()),
+                    tuple: identity.tuple.clone(),
+                    latch: identity.latch.clone(),
+                    interaction_type,
+                })]
+                .into_iter()
+            }
+            AnalyzedIdentity::Permutation(PermutationIdentity { left, right, .. })
+            | AnalyzedIdentity::PhantomPermutation(PhantomPermutationIdentity {
+                left,
+                right,
+                ..
+            }) => {
+                let id_left = id_counter;
+                let id_right = id_counter + 1;
+                id_counter += 2;
+                vec![
```

Comment:
Is there a way to make this less verbose? Like a function `construct_bus_interaction_pair()`?